stolon/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use tokio_socks::{tcp::Socks5Stream, TargetAddr};
+use tower::{Service, ServiceExt};
+
+pub use zebra_chain::{
+    parameters::Network, serialization::ZcashDeserialize, transaction::Transaction,
+};
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub async fn obtain_peer_addr(
+    tor_addr: SocketAddr,
+    network: Network,
+) -> Result<SocketAddr, BoxError> {
+    let seed_domain = match network {
+        Network::Mainnet => "mainnet.seeder.zfnd.org:8233",
+        Network::Testnet => "testnet.seeder.zfnd.org:18233",
+    };
+
+    tracing::debug!(?seed_domain, "using tor resolver to find peers");
+    // this API is pretty awkward, let's upstream some patches.
+    let mut peer_addr = match Socks5Stream::tor_resolve(tor_addr, seed_domain).await? {
+        TargetAddr::Ip(addr) => addr,
+        TargetAddr::Domain { .. } => return Err("resolving a domain gave a domain".into()),
+    };
+
+    peer_addr.set_port(match network {
+        Network::Mainnet => 8233,
+        Network::Testnet => 18233,
+    });
```

Comment:
Protocol Question:
Do seeders only supply Zcash instances on the default ports?

Style:
Maybe we should make this into a zebra-network function?
It seems like we've duplicated it a few times.

Reply:
Yeah, I'm not sure about this.  I don't know whether the seeders only supply Zcash instances that listen on the default ports, but on the other hand, because of the way that the DNS seeders work (DNS names can't resolve to a port), I don't see how we could learn the port except by guessing.  Maybe @gtank knows.

On the style question: yes, I think this probably makes sense to move into a function.  Maybe `Network::default_port(&self) -> u16`?

Reply:
> Yeah, I'm not sure about this. I don't know whether the seeders only supply Zcash instances that listen on the default ports, but on the other hand, because of the way that the DNS seeders work (DNS names can't resolve to a port), I don't see how we could learn the port except by guessing. Maybe @gtank knows.

I can't see any port filtering in [dnsseeder](https://github.com/zcashfoundation/dnsseeder) or [zcash-seeder](https://github.com/zcash/zcash-seeder).

Maybe it's not obvious, or maybe it was just left out?

> On the style question: yes, I think this probably makes sense to move into a function. Maybe `Network::default_port(&self) -> u16`?

Sounds good. Then we can replace the existing references to 8233 / 18233.

Reply:
The `default_port` method is now present in `zebra-network`. But I think that this discussion point was subsumed by https://github.com/ZcashFoundation/zebra/pull/1008#discussion_r484156174